components/server/ee/src/prebuilds/github-enterprise-app.ts

```diff
@@ -116,6 +116,15 @@ export class GitHubEnterpriseApp {
     ): Promise<StartPrebuildResult | undefined> {
         const span = TraceContext.startSpan("GitHubEnterpriseApp.handlePushHook", ctx);
         try {
+            const cloneURL = payload.repository.clone_url;
+            const projectAndOwner = await this.findProjectAndOwner(cloneURL, user);
+            if (projectAndOwner.project) {
```

Comment:
At the point of this call, how do we know that the project specified in the payload in fact belongs to the correct user? Is it possible to poison our data across projects here?

Mostly lacking context, so trying to understand better how the payload from the webhook is authhorized

Reply:
> At the point of this call, how do we know that the project specified in the payload in fact belongs to the correct user?

That's a good question. I don't know about webhook events from the GitHub App, but for GitLab, we look at the provided (user + project) token in order to make sure that the project specified in the payload actually belongs to the user who installed this webhook:

https://github.com/gitpod-io/gitpod/blob/e8fa6d20d2c18cdb0f02ff2cfd483388c5c5dac9/components/server/ee/src/prebuilds/gitlab-app.ts#L61-L97

We do something similar for GitHub Enterprise webhooks, where the secret token is not sent as part of the webhook payload, but is used to sign the webhook payload (which we can verify against the user's tokens):

https://github.com/gitpod-io/gitpod/blob/e8fa6d20d2c18cdb0f02ff2cfd483388c5c5dac9/components/server/ee/src/prebuilds/github-enterprise-app.ts#L89-L106

> Is it possible to poison our data across projects here?

I don't think it is possible to send forged webhook events to Gitpod.

However, your question does remind me of an issue where Pull Requests can come from forks, but trigger a Prebuild for the main repository, which can be problematic in some cases: https://github.com/gitpod-io/security/issues/26